### DISCUSSION TITLE
launcher configuration file for vtk_server.py

### DISCUSSION THREAD

**Discussion by guozhi72** [Kitware/trame#276] [General]
Hi,
Want to ask how to prepare configuration file to launch vtk_server.py at https://kitware.github.io/vtk-js/examples/RemoteView.html
I spend quite a lot of time to search and found out
1, I need to run python.exe some_folder/wslink/launcher.py laucher.config.
2, and I know on client side, java script, I need to give url using applition name instead of fixed port number.
But I don't really understand the example config file in launder.py:
1, for endpoint below, should I still give paraview

        "configuration": {
            "host" : "localhost",
            "port" : 8080,
            "endpoint": "paraview",                   // SessionManager Endpoint
            "content": "/.../www",                    // Optional: Directory shared over HTTP
            "proxy_file" : "/.../proxy-mapping.txt",  // Proxy-Mapping file for Apache
            "sessionURL" : "ws://${host}:${port}/ws", // ws url used by the client to connect to the started process
            "timeout" : 25,                           // Wait time in second after process start
2, for propeties: how many below should I give? I guess only I can skip for paraview related.

"properties" : {
            "vtkpython" : "/.../VTK/build/bin/vtkpython",
            "pvpython" : "/.../ParaView/build/bin/pvpython",
            "vtk_python_path": "/.../VTK/build/Wrapping/Python/vtk/web",
            "pv_python_path": "/.../ParaView/build/lib/site-packages/paraview/web",
            "plugins_path": "/.../ParaView/build/lib",
            "dataDir": "/.../path/to/data/directory"
        },
Based on my searched, I maybe able to debug launer.py to continue to try, but can somebody help of provide one configure example which can launch vtk_server.py.  I imagine it will be much less the the example in launcher.py.
Best

**Comment by jourdain** [comment]
Well if you were using trame-cookiecutter, you will get a docker bundle already done for you that take care of that part and more.

To answer your question the endpoint should still be "paraview" and the "properties" group can be empty. Those are just to reduce your command line sections. Also you should run it like `python -m wslink.launcher /path/to/config.json`.

**Comment by guozhi72** [comment]
Hi, Sebastian:

Thanks for your reply. 
I will try to use trame-cookiecutter later for sure. But it looks almost self contained. But I saw build.sh, and run.sh, so that implies I have to be in unix system. I imagine if I try on windows, anything can happened.

Redring using javascript and vtkweb, kitware has example. But if can add some example how to lauch vtk_server.py for multiple user that would be wonderful.

Another example is when I launch trame example, saw message "Note that for multi-users you need to use and configure a launcher." Totally have no idea what to do at the beginning. After spend a few weeks, on vtk.web, and trame, I guess I many need to run wsline/laucher.py mylauncer.config. But just still couldn't find an example. Have to try.

Best

**Reply by jourdain** [reply]
So the scripts are just command line for docker. On windows, you will just have to run those command by hand but they will work. 

For the multi user example, we highly recommend using docker. We provide a lot of doc on that subject but still not enough for many users. The default launcher config that get printed by default is already a great working example. 
Basically using the launcher without Apache is also very limited as it will use different port underneath rather than a single one like http or https.

Like I said, if you want to use a trame app for multiple users, you need something special. The cookiecutter give you that option. You also even have that [example](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/SingleFile) as well.

**Comment by guozhi72** [comment]
Hi, Sebastian:

Thanks for all the information. Trame did have big advantages. Though I did not test the deployment, I feel trame deployment documentation is solid.

But java script( for examle angular ) as UI, do visualization using vtk remote rendering also had it clearly advantage.
Do you have a docker example, which if I put vtk_server.py, then docker can handle mutiple user or at least multiple connection case.

**Reply by jourdain** [reply]
Just use the trame bundle. This is the same thing. It was just cleaned up and streamlined to align with trame philosophy.

**Comment by guozhi72** [comment]
do you mean I can use trame bundle to depoly : https://kitware.github.io/paraviewweb/examples/RemoteRenderer.html, this example?

**Reply by jourdain** [reply]
yes... You will need to change the client code to not use a fixed session url since that will be given to it by the launcher and you will have to adjust the command line or make it a python package for its server side.

You can find some info on how to use the [trame docker here](https://github.com/Kitware/trame/tree/master/docker)

**Comment by guozhi72** [comment]
for web_cone, https://kitware.github.io/paraviewweb/examples/RemoteRenderer.html. I hard code 1234 in client, and not using launcher, it works.  If use launcher, I got below error, I have the command I run, config file and client code at the end of letter( why I got  9999/paraview, should not be 8080/paraview)

        Failed to load resource: the server responded with a status of 404 (Not Found)
:9999/paraview/:1 

######## command I run
C:\dev_azure\server_render_angular_client\server\.env\Scripts\python.exe C:\dev_azure\server_render_angular_client\server\launcher.py launch.config 
####    launch.config :

  {
        // ===============================
        // General launcher configuration
        // ===============================

        "configuration": {
            "host" : "localhost",
            "port" : 8080,
            "endpoint": "paraview",                   // SessionManager Endpoint
            "sessionURL" : "ws://${host}:${port}/ws", // ws url used by the client to connect to the started process
            "timeout" : 25,                           // Wait time in second after process start
            "log_dir" : "C:\\dev_azure\\server_render_angular_client\\server\\log",             // Directory for log files
            "proxy_file" : "",
            "fields" : ["file", "host", "port"]
            },

        "resources" : [ { "host" : "localhost", "port_range" : [9001, 9003] } ],

        "properties" : {

        },

        "apps" : {
            "cone" : {
                "cmd" : [
                    "C:\\dev_azure\\server_render_angular_client\\server\\.env\\Scripts\\python.exe",
                    "C:\\dev_azure\\server_render_angular_client\\server\\vtk_server.py",
                    "--port",
                    "$port" ],
                "ready_line" : "Starting factory"
            }
        }
   }
        
    #### 222222    client code( pure java script

import 'normalize.css';

import SmartConnect from 'wslink/src/SmartConnect';

import RemoteRenderer from 'paraviewweb/src/NativeUI/Canvas/RemoteRenderer';
import SizeHelper from 'paraviewweb/src/Common/Misc/SizeHelper';
import ParaViewWebClient from 'paraviewweb/src/IO/WebSocket/ParaViewWebClient';

document.body.style.padding = '0';
document.body.style.margin = '0';

const divRenderer = document.createElement('div');
document.body.appendChild(divRenderer);

divRenderer.style.position = 'relative';
divRenderer.style.width = '100vw';
divRenderer.style.height = '100vh';
divRenderer.style.overflow = 'hidden';

// const config = { sessionURL: 'ws://localhost:1234/ws' };
const config = { application: 'cone' };
const smartConnect = SmartConnect.newInstance({ config });
smartConnect.onConnectionReady((connection) => {
  const pvwClient = ParaViewWebClient.createClient(connection, [
    'MouseHandler',
    'ViewPort',
    'ViewPortImageDelivery',
  ]);
  const renderer = new RemoteRenderer(pvwClient);
  renderer.setContainer(divRenderer);
  renderer.onImageReady(() => {
    console.log('We are good');
  });
  window.renderer = renderer;
  SizeHelper.onSizeChange(() => {
    renderer.resize();
  });
  SizeHelper.startListening();
});
smartConnect.connect();

**Comment by guozhi72** [comment]
I just could not find any example to make the web_con remote rendering work using launcher for multiple connection.

And since it is server-side rendering, multiple connection is mandatory. Surprisingly, could not find single example

**Comment by guozhi72** [comment]
If I check the browser, I got this. Of course 9999 cannot find.

Request URL:
http://localhost:9999/paraview/
Request Method:
POST
Status Code:
404 Not Found
Remote Address:
[::1]:9999
Referrer Policy:
strict-origin-when-cross-origin

**Reply by jourdain** [reply]
Yes that is expected since you have several servers. See my comment on Apache before. 
I guess you don't have much background with network and servers hence your struggle. 

What is the URL of the launcher (http://localhost:8080/paraview/) vs where does the client try to connect (http://localhost:9999/paraview/)? Why do you have a server on 9999? Looks like a development server... 

For a real deployment, everything should be served by the same server on the same port. If you want to override that default/production behavior, you can override the sessionManagerURL by using this URL `http://localhost:9999/?sessionManagerURL=http://localhost:8080/paraview`.  

The funny thing is once you will figure out the launcher, you will realized that is still not enough for what you need as you will have firewall issue. 

Out of curiosity. Do you know how to build that JavaScript code so it can then be served by a static http server like the launcher?

**Reply by guozhi72** [reply]
sorry, I miss to read this answer. Regarding your question, no, I don't really know. That is why I need an comprehensive but simple example. After that, we can do  a little more complex development.
Regarding the final deployment, we need to have more professinal person to do it. We even pay kitware before for consuting.
Out of curiosity. Do you know how to build that JavaScript code so it can then be served by a static http server like the launcher?

**Reply by guozhi72** [reply]
the java script client of webcone is on port 9999. 
############
What is the URL of the launcher (http://localhost:8080/paraview/) vs where does the client try to connect (http://localhost:9999/paraview/)? Why do you have a server on 9999? Looks like a development server...

**Reply by guozhi72** [reply]
I understand trame should be much simple to use. But using trame, we have to make the visualization stand alone. Otherwise, do purely client side rending is also easier to manage.

But at this moment, still want to try server side rendering, but visulization compoment, we want to embed into anguar frame work ( for client side rending, so simple to do embedding)

**Reply by jourdain** [reply]
First of all the vtk.js documentation can not cover, basic JavaScript or devOps. 
We provide the core information needed to understand the vtk part and how it would work with your environment (guides). 

If you are aiming to use a professional person for the deployment, why even bother? 
You can run the code locally on your own and it is working. So why do you need more? 
There will be no difference when used for several users on in a dev/single setup. 

BTW, I gave you the solution to your problem and it seems you didn't even noticed or tried it. 

Since you need everything spell out, here is the how-to for you:

1. Create Python virtual environment with VTK

```bash
python -m venv .venv
source .venv/bin/activate
pip install vtk wslink
```

2. Get the code example

```bash
curl https://raw.githubusercontent.com/Kitware/vtk-js/master/Sources/Rendering/Misc/RemoteView/example/vtk-server.py -O
curl https://kitware.github.io/vtk-js/examples/RemoteView/RemoteView.html -O
```

Then edit the HTML and remove the hard coded sessionURL by removing `,sessionURL:"ws://localhost:1234/ws"`. And rename it to index.html.

3. Setup the config file for the launcher.

file: launcher.conf
```
{
    "configuration": {
        "host" : "localhost",
        "port" : 8080,
        "content": ".",
        "endpoint": "paraview",
        "sessionURL" : "ws://${host}:${port}/ws",
        "timeout" : 25,
        "log_dir" : ".",
        "proxy_file" : "sessions.txt",
        "fields" : []
        },
    "resources" : [ { "host" : "localhost", "port_range" : [9001, 9100] } ],
    "properties" : {
    	"current_dir": "/Users/sebastien.jourdain/Desktop/test"
    },
    "apps" : {
        "cone" : {
            "cmd" : [
                "${current_dir}/.venv/bin/python",
                "${current_dir}/vtk-server.py",
                "--port",
                "$port"
            ],
            "ready_line" : "Starting factory"
        }
    }
}
```

5. Run the server

```bash
python -m wslink.launcher ./launcher.conf
```

Then open the browser at http://localhost:8080/

**Reply by guozhi72** [reply]
Hi, sebastien:

Thanks for you quick response and help. I was in VTK user forum 12 years ago, the help never came back so quick.

I will certainly try your suggestion. Regarding your question below:

Right now I can use angular component to do vtk remote rendering, it seems  I may continue do development. But I do doubts:
1, I even don't see (or don't know) how to do develoment for different component to  render different content( volume, cone, slice)
2, what I see the in laucher.config, I can have different app, volume/slice/cone, if I can run locally run npm start/serve to run client, and run laucher, then I see the way to move forward ( but now stucked)
3, right now,  unless in client code to do hard code for different application such as wslink:1234 for cone, 1235 for volume, I don't understand how to move forward ( and don't understand how static html can do this job).
4, but maybe, as you noticed, I just don't know.
5, From my current experimentnt,  ( the other two solution, pure client rendering or using trame, I see clearly and confident for forward develoment), but I don't see how to move forward even only for  local developme javascrip UI--vtk remote rendering.


##################
If you are aiming to use a professional person for the deployment, why even bother?
You can run the code locally on your own and it is working. So why do you need more?
There will be no difference when used for several users on in a dev/single setup.

**Reply by jourdain** [reply]
3. I gave you the solution already for not having any hard coded part in the code which can also work with static html. More information are in the launcher guide. There is nothing that I can do if you don't understand the solution provided in this thread and if you don't want to read the guides.

**Reply by guozhi72** [reply]
I will try your suggested. It looks clear,  I must be able to produce. After that, I will get back to you .
Thanks

**Reply by guozhi72** [reply]
Hi, @Sebastien:
I follow each step you directed, still could not make it works.

my folder structure is as
![Capture](https://github.com/Kitware/trame/assets/84816693/9badb907-40b3-43c8-bdab-02bdfb85c481)
I modify the html file and rename to index.html as shown:
![index](https://github.com/Kitware/trame/assets/84816693/3caa487f-69a7-470b-a63e-1b19da2a23ff)
my config is as:
![config](https://github.com/Kitware/trame/assets/84816693/4a52c92a-a343-4e61-8131-743503609dfe)

I run python launcher as 
![run](https://github.com/Kitware/trame/assets/84816693/7f00b8f1-f81f-4e60-b920-54614f0a014c)
I got errror message
![error](https://github.com/Kitware/trame/assets/84816693/9f661100-b39d-4b1d-ac4e-b0d9cb00d107)

the log is 

![log](https://github.com/Kitware/trame/assets/84816693/7049ebda-4f12-46cc-8835-88e3a0e6d6a0)

Good thing is I can see the individual server is launched, and get an idea. Can you see what is wrong?

**Reply by jourdain** [reply]
You still have `/` in the `cmd` section while you properly fixed the `current_dir` by using `\\`.
The problem was even shown in the error message.

**Reply by guozhi72** [reply]
thanks for your help, I change the path correctly.  But still got error.
here is the error:
![error](https://github.com/Kitware/trame/assets/84816693/880112f5-a85a-4fbf-900f-97e3d5388173)
Here is the log
![log](https://github.com/Kitware/trame/assets/84816693/45ed3fa1-5a98-43f7-8f85-17bff6f9396e)
Thanks for your help.

**Reply by jourdain** [reply]
The log you are sharing is not the useful one. You need to look at the one from the created session.

**Comment by jourdain** [comment]
Since you can't find the documentation I will list some links below.

- http://kitware.github.io/paraviewweb/docs/multi_user_setup.html
- http://kitware.github.io/paraviewweb/docs/launcher_api.html
- http://kitware.github.io/paraviewweb/docs/python_launcher.html
- http://kitware.github.io/paraviewweb/docs/launching_examples.html
- https://github.com/Kitware/paraviewweb-demo/
- https://github.com/Kitware/paraviewweb-demo/blob/master/pvw/launcher/config.json
- https://github.com/hydroframe/SandTank/blob/master/deploy/pvw/launcher/config.json

**Reply by guozhi72** [reply]
I mean an example for launching multiple web cone

**Reply by jourdain** [reply]
Well build that cone example for production and serve the static html using the launcher via the `"content": "/path/to/static-html/` property and that's it.

**Reply by jourdain** [reply]
If you understood what you were doing by reading the guides, you would not need an example.

**Comment by guozhi72** [comment]
Close
